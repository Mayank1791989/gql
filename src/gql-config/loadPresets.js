/* @flow */
/* @babel-flow-runtime-enable */
import path from 'path';
import importModule from 'gql-shared/importModule';
import { normalizePkgConfig } from './normalizePkg';
import { validateFragmentScopes } from 'gql-shared/FragmentScope';
import { type QueryPreset, type SchemaPreset, type PresetPkg } from './types';

import { reify, Type } from 'flow-runtime';

const GQL_MODULE_DIR = path.resolve(__dirname, '../');

export function loadQueryPresets(
  presets: ?Array<PresetPkg>,
  configPath: string,
): Array<QueryPreset> {
  return (presets || ['default']).map(presetPkg => {
    return loadPreset({
      preset: presetPkg,
      configPath,
      prefix: 'gql-query-preset',
      corePresets: [
        'gql-query-preset-relay',
        'gql-query-preset-relay-modern',
        'gql-query-preset-apollo',
        'gql-query-preset-default',
      ],
      validate(presetConfig) {
        const QueryPresetType = (reify: Type<QueryPreset>);
        // validate type
        QueryPresetType.assert(presetConfig);

        // validate fragmentScopes if present
        if (!presetConfig.fragmentScopes) {
          return;
        }
        const error = validateFragmentScopes(presetConfig.fragmentScopes);
        if (error) {
          throw error;
        }
      },
    });
  });
}

export function loadSchemaPresets(
  presets: ?Array<PresetPkg>,
  configPath: string,
): Array<SchemaPreset> {
  return (presets || ['default']).map(presetPkg => {
    return loadPreset({
      preset: presetPkg,
      configPath,
      prefix: 'gql-schema-preset',
      corePresets: ['gql-schema-preset-default'],
      validate(presetModule) {
        const SchemaPresetType = (reify: Type<SchemaPreset>);
        SchemaPresetType.assert(presetModule);
      },
    });
  });
}

function loadPreset(params: {
  preset: PresetPkg,
  prefix: string,
  validate: Function,
  corePresets: Array<string>,
  configPath: string,
}) {
  const [presetPkg, options] = normalizePkgConfig(params.prefix, params.preset);

  const [pkg, dir] = params.corePresets.includes(presetPkg)
    ? [`./${presetPkg}`, GQL_MODULE_DIR]
    : [presetPkg, params.configPath];

  try {
    const presetModule = importModule(pkg, dir);

    if (typeof presetModule !== 'function') {
      throw new Error(
        `PRESET_PKG_INVALID: expected preset "${presetPkg}" to export "function" but found "${typeof presetModule}"`,
      );
    }

    const presetConfig = presetModule(options);
    presetConfig.name = pkg;

    // validate preset
    params.validate(presetConfig);

    return presetConfig;
  } catch (err) {
    if (err.code === 'MODULE_NOT_FOUND') {
      throw new Error(
        `PRESET_PKG_NOT_FOUND: Preset '${pkg}' not found relative to '${dir}'`,
      );
    }

    if (err.name === 'RuntimeTypeError') {
      throw new Error(
        `PRESET_CONFIG_INVALID: There are errors in preset '${pkg}'\n\n${
          err.message
        }`,
      );
    }

    throw err;
  }
}
