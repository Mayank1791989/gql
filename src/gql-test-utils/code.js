/* @flow */
import { dedent } from 'dentist';
import splitLines from 'gql-shared/splitLines';

export default function code(text: string) {
  const sourceText = dedent(text);
  const lines = splitLines(sourceText);
  let position = null;
  lines.forEach((line, index) => {
    const match = line.match(/--\^/);
    if (match) {
      position = {
        // $FlowDisableNextLine
        column: match.index + 3,
        line: index,
      };
    }
  });

  if (!position) {
    throw new Error('Missing --^ in source');
  }

  return {
    sourceText,
    position,
  };
}
