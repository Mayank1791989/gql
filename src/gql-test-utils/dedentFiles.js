/* @flow */
import { dedent } from 'dentist';

type FilesMap = { [path: string]: string };

export function dedentFiles(files?: FilesMap): FilesMap {
  if (!files) {
    return {};
  }
  return Object.keys(files).reduce((acc, path) => {
    acc[path] = dedent(files[path]);
    return acc;
  }, {});
}
