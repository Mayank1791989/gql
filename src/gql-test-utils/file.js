/* @flow */
import os from 'os';
import objectHash from 'object-hash';
import { writeFileSync } from 'gql-shared/fs-utils';
import { execSync } from 'child_process';
import path from 'path';

export const TEST_TEMP_DIR = path.join(os.tmpdir(), 'gql-test');

export function createTempFiles(files: { [file: string]: string }): string {
  // deterministic dirName [as this path used in test cases output (see generated snapshot)]
  const dir = path.join(TEST_TEMP_DIR, objectHash(files));
  Object.keys(files).forEach(file => {
    const content = files[file];
    writeFileSync(path.join(dir, file), content);
  });
  return dir;
}

export function removeTestTempDir() {
  execSync(`rm -rf ${TEST_TEMP_DIR}`);
}
