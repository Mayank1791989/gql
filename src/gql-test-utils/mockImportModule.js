/* @flow */
export default function mockImportModule(
  moduleName: string,
  mockFactory: () => $FixMe,
) {
  jest.doMock('import-from', () => {
    return function importFrom(dir, moduleID) {
      if (moduleID === moduleName) {
        return mockFactory();
      }
      const importFromActual = jest.requireActual('import-from');
      return importFromActual(dir, moduleID);
    };
  });
}
