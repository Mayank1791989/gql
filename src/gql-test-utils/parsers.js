/* @flow */
import SchemaParser from 'gql-schema-parser-default';
import EmbeddedQueryParser from 'gql-query-parser-embedded-queries';
import QueryParser from 'gql-query-parser-default';
import QueryPresetRelay from 'gql-query-preset-relay';

export function relayQLParser() {
  return new EmbeddedQueryParser({
    start: 'Relay\\.QL`',
    end: '`',
    allowFragmentWithoutName: true,
    allowFragmentInterpolation: true,
  });
}

export function apolloParser() {
  return new EmbeddedQueryParser({
    start: 'gql`',
    end: '`',
    allowDocumentInterpolation: true,
  });
}

export const relayExtendSchema = QueryPresetRelay().extendSchema;

export function queryParser() {
  return new QueryParser();
}

export function customParser() {
  return new EmbeddedQueryParser({
    start: '"""',
    end: '"""',
  });
}

export function embeddedQueryParser({
  start,
  end,
}: {
  start: string,
  end: string,
}) {
  return new EmbeddedQueryParser({
    start,
    end,
  });
}

export function schemaParser() {
  return new SchemaParser();
}
