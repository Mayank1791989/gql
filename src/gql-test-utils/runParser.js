/* @flow */
import { type IParser } from 'gql-shared/types';
import MultilineCharacterStream from 'gql-shared/MultilineCharacterStream';

export default function runParser({
  sourceText,
  parser,
}: {
  sourceText: string,
  parser: IParser,
}) {
  const state = parser.startState();
  const stream = new MultilineCharacterStream(sourceText);
  const result = [];
  while (!stream.eof()) {
    const style = parser.token(stream, state);
    result.push({
      start: stream.getStartOfToken(),
      end: stream.getCurrentPosition(),
      text: stream.current(),
      style,
    });
  }

  return result;
}
