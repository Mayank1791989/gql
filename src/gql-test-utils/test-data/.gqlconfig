{
  schema: {
    files: "schema/**/*.gql",
    graphQLOptions: {
      commentDescriptions: true,
    },
  }
}
