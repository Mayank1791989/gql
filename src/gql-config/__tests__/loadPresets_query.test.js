/* @flow */
import { loadQueryPresets } from '../loadPresets';
import { createTempFiles } from 'gql-test-utils/file';

describe('load core-presets:', () => {
  [
    'gql-query-preset-default',
    'gql-query-preset-relay',
    'gql-query-preset-apollo',
  ].forEach(preset => {
    test(`using full package name =${preset}`, () => {
      expect(loadQueryPresets([preset], '')).toBeDefined();
    });
  });

  ['default', 'relay', 'apollo'].forEach(preset => {
    test(`using short package name =${preset}`, () => {
      expect(loadQueryPresets([preset], '')).toBeDefined();
    });
  });
});

describe('missing-preset-pkg', () => {
  [
    'gql-query-preset-custom',
    './custom-preset',
    '@scope/custom',
    'custom_node_module',
  ].forEach(preset => {
    test(`should throw when preset=${preset}`, () => {
      expect(() =>
        loadQueryPresets([preset], ''),
      ).toThrowErrorMatchingSnapshot();
    });
  });
});

describe('custom-preset', () => {
  test('load preset if it is defined in local file', () => {
    const rootPath = createTempFiles({
      'custom-preset': `
        /* this is invalid preset file */
        module.exports = function customPreset() {
          return {
            parserOptions: { },
            validate: {
              rules: {
                someCustomRule : function someCustomRule(context) { },
              },
              config: {
                someCustomRule: 'error',
              },
            }
          };
        };
      `,
    });

    expect(loadQueryPresets(['./custom-preset'], rootPath))
      .toMatchInlineSnapshot(`
Array [
  Object {
    "name": "./custom-preset",
    "preset": Object {
      "parserOptions": Object {},
      "validate": Object {
        "config": Object {
          "someCustomRule": "error",
        },
        "rules": Object {
          "someCustomRule": [Function],
        },
      },
    },
  },
]
`);
  });

  describe('load preset if it is node_module', () => {
    const rootPath = createTempFiles({
      'node_modules/gql-query-preset-custom/package.json': `
        {
          "name": "gql-query-preset-custom",
          "version": "1.0"
        }
      `,
      'node_modules/gql-query-preset-custom/index.js': `
        module.exports = function customPreset() {
          return {
            parserOptions: { },
            validate: {
              rules: {
                someCustomRule: function someCustomRule(context) { },
              },
              config: {
                someCustomRule: 'error',
              },
            }
          };
        };
      `,
    });

    test('using full package name', () => {
      expect(loadQueryPresets(['gql-query-preset-custom'], rootPath))
        .toMatchInlineSnapshot(`
Array [
  Object {
    "name": "gql-query-preset-custom",
    "preset": Object {
      "parserOptions": Object {},
      "validate": Object {
        "config": Object {
          "someCustomRule": "error",
        },
        "rules": Object {
          "someCustomRule": [Function],
        },
      },
    },
  },
]
`);
    });

    test('using short package name', () => {
      expect(loadQueryPresets(['custom'], rootPath)).toMatchInlineSnapshot(`
Array [
  Object {
    "name": "gql-query-preset-custom",
    "preset": Object {
      "parserOptions": Object {},
      "validate": Object {
        "config": Object {
          "someCustomRule": "error",
        },
        "rules": Object {
          "someCustomRule": [Function],
        },
      },
    },
  },
]
`);
    });
  });
});

test('Validation: throw if preset doesnt export function', () => {
  const rootPath = createTempFiles({
    'custom-preset': `
        /* this is invalid preset file */
        module.exports = {
          parserOptions: { }
        };
      `,
  });

  expect(() => {
    loadQueryPresets(['./custom-preset'], rootPath);
  }).toThrowErrorMatchingInlineSnapshot(
    // eslint-disable-next-line playlyfe/babel-quotes
    `"PRESET_PKG_INVALID: expected preset \\"./custom-preset\\" to export \\"function\\" but found \\"object\\""`,
  );
});

test('Validation: throw if preset config invalid', () => {
  const rootPath = createTempFiles({
    'custom-preset': `
        /* this is invalid preset file */
        module.exports = function () {
          return {
            parserOptions: { },
            validate: {},
          };
        };
      `,
  });

  expect(() => {
    loadQueryPresets(['./custom-preset'], rootPath);
  }).toThrowErrorMatchingInlineSnapshot(`
"PRESET_CONFIG_INVALID: There are errors in preset './custom-preset'

QueryPreset.validate.config must be an object

Expected: {
  [ruleName: string]: \\"off\\" | \\"warn\\" | \\"error\\";
}

Actual Value: undefined

Actual Type: void
"
`);
});
