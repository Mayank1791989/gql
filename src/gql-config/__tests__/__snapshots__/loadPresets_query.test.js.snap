// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validation: throw if preset config invalid 1`] = `
"PRESET_CONFIG_INVALID: There are errors in preset './custom-preset'

QueryPreset.validate.config must be an object

Expected: {
  [ruleName: string]: \\"off\\" | \\"warn\\" | \\"error\\";
}

Actual Value: undefined

Actual Type: void
"
`;

exports[`Validation: throw if preset doesnt export function 1`] = `"PRESET_PKG_INVALID: expected preset \\"./custom-preset\\" to export \\"function\\" but found \\"object\\""`;

exports[`custom-preset load preset if it is defined in local file 1`] = `
Array [
  Object {
    "name": "./custom-preset",
    "parserOptions": Object {},
    "validate": Object {
      "config": Object {
        "someCustomRule": "error",
      },
      "rules": Object {
        "someCustomRule": [Function],
      },
    },
  },
]
`;

exports[`custom-preset load preset if it is node_module using full package name 1`] = `
Array [
  Object {
    "name": "gql-query-preset-custom",
    "parserOptions": Object {},
    "validate": Object {
      "config": Object {
        "someCustomRule": "error",
      },
      "rules": Object {
        "someCustomRule": [Function],
      },
    },
  },
]
`;

exports[`custom-preset load preset if it is node_module using short package name 1`] = `
Array [
  Object {
    "name": "gql-query-preset-custom",
    "parserOptions": Object {},
    "validate": Object {
      "config": Object {
        "someCustomRule": "error",
      },
      "rules": Object {
        "someCustomRule": [Function],
      },
    },
  },
]
`;

exports[`missing-preset-pkg should throw when preset=./custom-preset 1`] = `"PRESET_PKG_NOT_FOUND: Preset './custom-preset' not found relative to ''"`;

exports[`missing-preset-pkg should throw when preset=@scope/custom 1`] = `"PRESET_PKG_NOT_FOUND: Preset '@scope/gql-query-preset-custom' not found relative to ''"`;

exports[`missing-preset-pkg should throw when preset=custom_node_module 1`] = `"PRESET_PKG_NOT_FOUND: Preset 'gql-query-preset-custom_node_module' not found relative to ''"`;

exports[`missing-preset-pkg should throw when preset=gql-query-preset-custom 1`] = `"PRESET_PKG_NOT_FOUND: Preset 'gql-query-preset-custom' not found relative to ''"`;
